### PR TITLE
Conditionally enable immediate full sync, based on value of constant

### PR DESCRIPTION
**Pull request by nickdaugherty**
## Description

JP 8.0+ includes a new full sync strategy - immediate.

This adds a filter that enables the new strategy, based on the value of the `VIP_JETPACK_FULL_SYNC_IMMEDIATELY` constant, which will be used to do early testing and a gradual rollout of this feature.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [ ] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [x] This change has relevant documentation additions / updates (if applicable).

## Steps to Test

TBD

### CODE REVIEW

File: vip-jetpack/vip-jetpack.php [hunk 1]
```diff
@@ -218,3 +218,20 @@ function wpcom_vip_disable_jetpack_email_no_recaptcha( $is_enabled ) {
 	return defined( 'RECAPTCHA_PUBLIC_KEY' ) && defined( 'RECAPTCHA_PRIVATE_KEY' );
 }
 add_filter( 'sharing_services_email', 'wpcom_vip_disable_jetpack_email_no_recaptcha', PHP_INT_MAX );
+
+/**
+ * Enable the new Full Sync method on sites with the VIP_JETPACK_FULL_SYNC_IMMEDIATELY constant
+ */
+add_filter( 'jetpack_sync_modules', function( $modules ) {
+	if ( ! defined( 'VIP_JETPACK_FULL_SYNC_IMMEDIATELY' ) || ! VIP_JETPACK_FULL_SYNC_IMMEDIATELY ) {
```
Comment (mjangda) [inline vip-jetpack/vip-jetpack.php]: Minor: can we reverse this and move the loop inside? I find the multiple `!` checks can be hard to read and somewhat fragile.

```suggestion
	if ( defined( 'VIP_JETPACK_FULL_SYNC_IMMEDIATELY' ) && true === VIP_JETPACK_FULL_SYNC_IMMEDIATELY ) {
```

Comment (nickdaugherty) [reply vip-jetpack/vip-jetpack.php]: Good call, done in 9c4f8034

**Comment by mjangda** [comment]
```
Fatal error:  Uncaught Error: Class 'Automattic\Jetpack\Sync\Modules\Full_Sync_Immediately' not found in wp-content/mu-plugins/jetpack/vendor/automattic/jetpack-sync/src/Modules.php:167
```

File: vip-jetpack/vip-jetpack.php [hunk 2]
```diff
@@ -218,3 +218,18 @@ function wpcom_vip_disable_jetpack_email_no_recaptcha( $is_enabled ) {
 	return defined( 'RECAPTCHA_PUBLIC_KEY' ) && defined( 'RECAPTCHA_PRIVATE_KEY' );
 }
 add_filter( 'sharing_services_email', 'wpcom_vip_disable_jetpack_email_no_recaptcha', PHP_INT_MAX );
+
+/**
+ * Enable the new Full Sync method on sites with the VIP_JETPACK_FULL_SYNC_IMMEDIATELY constant
+ */
+add_filter( 'jetpack_sync_modules', function( $modules ) {
+	if ( defined( 'VIP_JETPACK_FULL_SYNC_IMMEDIATELY' ) && true === VIP_JETPACK_FULL_SYNC_IMMEDIATELY ) {
```
Comment (mjangda) [inline vip-jetpack/vip-jetpack.php]: Need a `class_exists` check here for `Automattic\Jetpack\Sync\Modules\Full_Sync_Immediately`

**Comment by mjangda** [comment]
Added `class_exists` check and tests.

File: vip-jetpack/vip-jetpack.php [hunk 3]
```diff
@@ -218,3 +218,22 @@ function wpcom_vip_disable_jetpack_email_no_recaptcha( $is_enabled ) {
...
+ * Enable the new Full Sync method on sites with the VIP_JETPACK_FULL_SYNC_IMMEDIATELY constant
+ */
+add_filter( 'jetpack_sync_modules', function( $modules ) {
+	if ( ! class_exists( 'Automattic\\Jetpack\\Sync\\Modules\\Full_Sync_Immediately' ) ) {
+		return $modules;
+	}
+
+	if ( defined( 'VIP_JETPACK_FULL_SYNC_IMMEDIATELY' ) && true === VIP_JETPACK_FULL_SYNC_IMMEDIATELY ) {
+		foreach ( $modules as $key => $module ) {
+			if ( in_array( $module, [ 'Automattic\\Jetpack\\Sync\\Modules\\Full_Sync', 'Jetpack_Sync_Modules_Full_Sync' ], true ) ) {
```
Comment (joshbetz) [inline vip-jetpack/vip-jetpack.php]:
```suggestion
			// Replace Jetpack_Sync_Modules_Full_Sync or Full_Sync with the new module
			if ( in_array( $module, [ 'Automattic\\Jetpack\\Sync\\Modules\\Full_Sync', 'Jetpack_Sync_Modules_Full_Sync' ], true ) ) {
```

Comment (joshbetz) [reply vip-jetpack/vip-jetpack.php]: The formatting of the diff looks weird, but this just adds a comment above the `if ( in_array( ...` line.

**Comment by mjangda** [comment]
r699-stacks

**Comment by nickdaugherty** [comment]
Thanks for catching that issue and adding tests @mjangda !

**Pull request merged**